ASoC: SOF: Reduce the use of sdev->bar[] directly in the core

ujfalusi opened the pull request:
The sdev->bar[] is platform and IO dependent construct in the common snd_sof_dev struct and by directly using it makes the core to be not portable to platforms where the DSP access is other than through mapped memory.

The core should not need to know the way how the DSP is accessed.

ranj063 commented on sound/soc/sof/ipc.c:
```diff
@@ -703,12 +703,12 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_control *scontrol,
 
 		else
-			snd_sof_dsp_block_read(sdev, sdev->mmio_bar,
+			snd_sof_dsp_block_read(sdev, SOF_FW_BLK_TYPE_IRAM,
```
now we have to check the return value here and all the other places iisnt it?

ujfalusi replied:
We don't need to check for the TYPE_IRAM mapping validity as it has been already tested earlier, but good point, let me check it again if we need to check here again.

ujfalusi replied:
I will add dev_err_once() to not flood dmesg. The worst thing that can happen if the IRAM is not mapped is that we never do real read/write.

Before this patch however the  sdev->mmio_bar was not verified:
if sdev->mmio_bar was not initialized then we would always access to sdev->bar[0] and if that is not initialized then we would try to access to address 0+offset.
If the sdev->mmio_bar could be also set to any arbitrary number causing out of bound accesses.
One would be able to craft a firmware image which would pass malicious bar ID for a given platform causing kernel crash as well.

lyakh commented on sound/soc/sof/loader.c:
```diff
@@ -359,8 +360,9 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
-static void sof_get_windows(struct snd_sof_dev *sdev, int bar)
+static void sof_get_windows(struct snd_sof_dev *sdev)
 {
+	int bar = snd_sof_dsp_get_bar_index(sdev, SOF_FW_BLK_TYPE_SRAM);
```
this seems to exactly undo the previous patch in the series with the only exception of not re-adding the error handling. Can we simply remove the previous patch?

ujfalusi replied:
Good point, let me see how it will change the code, it should be straight forward to drop that patch.

ujfalusi replied:
Done

lyakh commented on sound/soc/sof/loader.c:
```diff
@@ -596,8 +589,13 @@ int snd_sof_parse_module_memcpy(struct snd_sof_dev *sdev,
+		ret = snd_sof_dsp_block_write(sdev, block->type, offset,
+					      block + 1, block->size);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: no BAR mapping for block type 0x%x\n",
```
shouldn't the error message be updated now too?

ujfalusi replied:
I have already changed it to "error: write to block type 0x%x failed\n"

kv2019i reviewed (commented):
Looks good to me @ujfalusi Once you update with the changes already discussed, this seems ready to go.

ujfalusi commented:
Hi,

Changes since v1:
- error prints revisited to match with the change
- as suggested by @lyakh I have removed the intermediate commit which was passing the bar as parameter
- Check for error as @ranj063 suggested and use dev_err_once()
- Added two patch (first two) for IMX to make sure that a deliberately modified firmware can not cause a kernel crash. (@dbaluta, can you check if it is correct?)

dbaluta commented:
@ujfalusi let me check this this week.

ujfalusi commented:
@dbaluta, thanks. If the crafted firmware would contain SOF_FW_BLK_TYPE_DRAM then not too healthy things might happen, it is better to make sure that only IRAM and SRAM type returns valid bar index.
It might be even possible to access exact addresses with a correct offset (0+offset == address).

dbaluta commented:
@ujfalusi I see your point. 

Although SOF_FW_BLK_TYPE_DRAM might be a valid bar in the future. 

But another interesting issue comes now.

What prevents an attacker to craft a firmware with correct BAR regions but with a bigger offset? It can crash the kernel right? I don't think we have a check for that.

ranj063 commented on sound/soc/sof/debug.c:
```diff
@@ -588,6 +588,21 @@ int snd_sof_debugfs_io_item(struct snd_sof_dev *sdev,
 }
 EXPORT_SYMBOL_GPL(snd_sof_debugfs_io_item);
 
+int snd_sof_debugfs_add_region_item_iomem(struct snd_sof_dev *sdev,
```
only minor. in the comit message " generic API...for generic DSP regions"? maybe a bit too generic :)

ujfalusi replied:
@ranj063, yes, it is too generic, let me reword it ;)

ujfalusi commented:
@dbaluta, validating the offset for a window is a bit tricky, we would need another set of callbacks defined for them.
I think the first two patch for imx is a valid one, it makes things a bit easier to follow.

I will have a followup series which will remove the bar use in core even from the helper functions. I still have few grey areas, but IMX might be able to get rid of using bar and have only the needed __iomem* .
Or not, I need to see how it will look.

Can I keep the patches or should I drop them?

lyakh commented on sound/soc/sof/imx/imx8.c:
```diff
@@ -364,7 +364,11 @@ static int imx8_remove(struct snd_sof_dev *sdev)
-	return type;
+	/* Only IRAM and SRAM bars are valid */
+	if (type == SOF_FW_BLK_TYPE_IRAM || type == SOF_FW_BLK_TYPE_SRAM)
+		return type;
```
a *soft* comment - maybe a `switch` would look better here and in imx8m too of course, but not a reason for a new version, only if you have to re-spin it for some other reason.

dbaluta replied:
@ujfalusi lets go with the switch as in the future we will also add DRAM bar. Other than that it looks good to me. 

We can merge this after the change.

ujfalusi replied:
@lyakh, @dbaluta, good point, I'll update it and push a new version (after lunch)

dbaluta commented:
@ujfalusi great work! Let me test this changes today and then hit the approve button.

dbaluta commented:
@ujfalusi basic tests on IMX look good!
